pkg/distributor/ha_tracker.go

```diff
@@ -212,10 +212,47 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
 		t.client = client
 	}
 
-	t.Service = services.NewBasicService(nil, t.loop, nil)
+	t.Service = services.NewBasicService(t.syncHATrackerStateOnStart, t.loop, nil)
 	return t, nil
 }
 
+func (h *haTracker) syncHATrackerStateOnStart(ctx context.Context) error {
+	if !h.cfg.EnableHATracker {
+		return nil
+	}
+	// haTracker holds HATrackerConfig with the prefix.
+	keys, err := h.client.List(ctx, "")
+	if err != nil {
+		return err
```

Comment:
After running the local setup in [microservices mode ](https://github.com/grafana/mimir/tree/main/development/mimir-microservices-mode), I noticed that this particular piece of code makes the ruler, and queriers fail because of connection issue to consul. 

**Why ?** 
This happens because both `querier` and `ruler` have `distributor` as a dependency in the code. Which means that they initialize the service in both scenarios. 

**Before - init cache**
Before the initialization it was also  failing but when the distributor was in Running state which didn't make the ruler, querier initialization fail.

**After - Init Cache**
Now, the error occurs earlier when the distributor is in the Starting state, causing both ruler and querier to fail.

**How we can solve it ?**
1. Override the mimir.yaml for ruler, and querier to not initialize the ha_tracker : ✅ 
2. Provide connection to consul, which creates confusion since we can see in metrics the ha_tracker from both ruler and querier, thats why I rejected it ❌